# This is my Lab 2

In [2]:
import warnings
warnings.simplefilter('ignore')

import pandas as pd

df = pd.read_csv('s3://turner-ia340-2025-fall/Crime_Data_from_2020_to_Present.csv') # load the data from s3
df.head() #show the top 5 rows

DR_NO               Date Rptd                DATE OCC  TIME OCC  AREA  \
0  211507896  04/11/2021 12:00:00 AM  11/07/2020 12:00:00 AM       845    15   
1  201516622  10/21/2020 12:00:00 AM  10/18/2020 12:00:00 AM      1845    15   
2  240913563  12/10/2024 12:00:00 AM  10/30/2020 12:00:00 AM      1240     9   
3  210704711  12/24/2020 12:00:00 AM  12/24/2020 12:00:00 AM      1310     7   
4  201418201  10/03/2020 12:00:00 AM  09/29/2020 12:00:00 AM      1830    14   

     AREA NAME  Rpt Dist No  Part 1-2  Crm Cd  \
0  N Hollywood         1502         2     354   
1  N Hollywood         1521         1     230   
2     Van Nuys          933         2     354   
3     Wilshire          782         1     331   
4      Pacific         1454         1     420   

                                         Crm Cd Desc  ... Status  Status Desc  \
0                                  THEFT OF IDENTITY  ...     IC  Invest Cont   
1     ASSAULT WITH DEADLY WEAPON, AGGRAVATED ASSAULT  ...     IC  Invest Cont   
2                                  THEFT OF IDENTITY  ...     IC  Invest Cont   
3  THEFT FROM MOTOR VEHICLE - GRAND ($950.01 AND ...  ...     IC  Invest Cont   
4    THEFT FROM MOTOR VEHICLE - PETTY ($950 & UNDER)  ...     IC  Invest Cont   

  Crm Cd 1 Crm Cd 2  Crm Cd 3 Crm Cd 4  \
0    354.0      NaN       NaN      NaN   
1    230.0      NaN       NaN      NaN   
2    354.0      NaN       NaN      NaN   
3    331.0      NaN       NaN      NaN   
4    420.0      NaN       NaN      NaN   

                                   LOCATION Cross Street      LAT       LON  
0   7800    BEEMAN                       AV          NaN  34.2124 -118.4092  
1           ATOLL                        AV     N  GAULT  34.1993 -118.4203  
2  14600    SYLVAN                       ST          NaN  34.1847 -118.4509  
3   6000    COMEY                        AV          NaN  34.0339 -118.3747  
4                   4700    LA VILLA MARINA          NaN  33.9813 -118.4350  

[5 rows x 28 columns]

In [3]:
df.shape #show the number of rows and columns 

(1004991, 28)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1004991 entries, 0 to 1004990
Data columns (total 28 columns):
 #   Column          Non-Null Count    Dtype  
---  ------          --------------    -----  
 0   DR_NO           1004991 non-null  int64  
 1   Date Rptd       1004991 non-null  object 
 2   DATE OCC        1004991 non-null  object 
 3   TIME OCC        1004991 non-null  int64  
 4   AREA            1004991 non-null  int64  
 5   AREA NAME       1004991 non-null  object 
 6   Rpt Dist No     1004991 non-null  int64  
 7   Part 1-2        1004991 non-null  int64  
 8   Crm Cd          1004991 non-null  int64  
 9   Crm Cd Desc     1004991 non-null  object 
 10  Mocodes         853372 non-null   object 
 11  Vict Age        1004991 non-null  int64  
 12  Vict Sex        860347 non-null   object 
 13  Vict Descent    860335 non-null   object 
 14  Premis Cd       1004975 non-null  float64
 15  Premis Desc     1004403 non-null  object 
 16  Weapon Used Cd  327247 non-null   fl

In [5]:
df.describe()

DR_NO      TIME OCC          AREA   Rpt Dist No      Part 1-2  \
count  1.004991e+06  1.004991e+06  1.004991e+06  1.004991e+06  1.004991e+06   
mean   2.202215e+08  1.339900e+03  1.069174e+01  1.115633e+03  1.400348e+00   
std    1.319718e+07  6.510613e+02  6.110255e+00  6.111605e+02  4.899691e-01   
min    8.170000e+02  1.000000e+00  1.000000e+00  1.010000e+02  1.000000e+00   
25%    2.106169e+08  9.000000e+02  5.000000e+00  5.870000e+02  1.000000e+00   
50%    2.209159e+08  1.420000e+03  1.100000e+01  1.139000e+03  1.000000e+00   
75%    2.311103e+08  1.900000e+03  1.600000e+01  1.613000e+03  2.000000e+00   
max    2.521041e+08  2.359000e+03  2.100000e+01  2.199000e+03  2.000000e+00   

             Crm Cd      Vict Age     Premis Cd  Weapon Used Cd      Crm Cd 1  \
count  1.004991e+06  1.004991e+06  1.004975e+06   327247.000000  1.004980e+06   
mean   5.001568e+02  2.891706e+01  3.056201e+02      363.955300  4.999174e+02   
std    2.052731e+02  2.199272e+01  2.193021e+02      123.734528  2.050736e+02   
min    1.100000e+02 -4.000000e+00  1.010000e+02      101.000000  1.100000e+02   
25%    3.310000e+02  0.000000e+00  1.010000e+02      311.000000  3.310000e+02   
50%    4.420000e+02  3.000000e+01  2.030000e+02      400.000000  4.420000e+02   
75%    6.260000e+02  4.400000e+01  5.010000e+02      400.000000  6.260000e+02   
max    9.560000e+02  1.200000e+02  9.760000e+02      516.000000  9.560000e+02   

           Crm Cd 2     Crm Cd 3   Crm Cd 4           LAT           LON  
count  69160.000000  2314.000000   64.00000  1.004991e+06  1.004991e+06  
mean     958.101258   984.015990  991.21875  3.399821e+01 -1.180909e+02  
std      110.354348    52.350982   27.06985  1.610713e+00  5.582386e+00  
min      210.000000   310.000000  821.00000  0.000000e+00 -1.186676e+02  
25%      998.000000   998.000000  998.00000  3.401470e+01 -1.184305e+02  
50%      998.000000   998.000000  998.00000  3.405890e+01 -1.183225e+02  
75%      998.000000   998.000000  998.00000  3.416490e+01 -1.182739e+02  
max      999.000000   999.000000  999.00000  3.433430e+01  0.000000e+00

I have 1004991 rows and 28 columns. About half are numeric and the other half are nonnumeric. It has very precise geolocation data.

In [6]:
#rename columns
df = df.rename(columns={"Crm Cd Desc": "Crime Description"})
df.head()

DR_NO               Date Rptd                DATE OCC  TIME OCC  AREA  \
0  211507896  04/11/2021 12:00:00 AM  11/07/2020 12:00:00 AM       845    15   
1  201516622  10/21/2020 12:00:00 AM  10/18/2020 12:00:00 AM      1845    15   
2  240913563  12/10/2024 12:00:00 AM  10/30/2020 12:00:00 AM      1240     9   
3  210704711  12/24/2020 12:00:00 AM  12/24/2020 12:00:00 AM      1310     7   
4  201418201  10/03/2020 12:00:00 AM  09/29/2020 12:00:00 AM      1830    14   

     AREA NAME  Rpt Dist No  Part 1-2  Crm Cd  \
0  N Hollywood         1502         2     354   
1  N Hollywood         1521         1     230   
2     Van Nuys          933         2     354   
3     Wilshire          782         1     331   
4      Pacific         1454         1     420   

                                   Crime Description  ... Status  Status Desc  \
0                                  THEFT OF IDENTITY  ...     IC  Invest Cont   
1     ASSAULT WITH DEADLY WEAPON, AGGRAVATED ASSAULT  ...     IC  Invest Cont   
2                                  THEFT OF IDENTITY  ...     IC  Invest Cont   
3  THEFT FROM MOTOR VEHICLE - GRAND ($950.01 AND ...  ...     IC  Invest Cont   
4    THEFT FROM MOTOR VEHICLE - PETTY ($950 & UNDER)  ...     IC  Invest Cont   

  Crm Cd 1 Crm Cd 2  Crm Cd 3 Crm Cd 4  \
0    354.0      NaN       NaN      NaN   
1    230.0      NaN       NaN      NaN   
2    354.0      NaN       NaN      NaN   
3    331.0      NaN       NaN      NaN   
4    420.0      NaN       NaN      NaN   

                                   LOCATION Cross Street      LAT       LON  
0   7800    BEEMAN                       AV          NaN  34.2124 -118.4092  
1           ATOLL                        AV     N  GAULT  34.1993 -118.4203  
2  14600    SYLVAN                       ST          NaN  34.1847 -118.4509  
3   6000    COMEY                        AV          NaN  34.0339 -118.3747  
4                   4700    LA VILLA MARINA          NaN  33.9813 -118.4350  

[5 rows x 28 columns]

In [7]:
df.isnull().sum() #show the number of rows containing null value

DR_NO                      0
Date Rptd                  0
DATE OCC                   0
TIME OCC                   0
AREA                       0
AREA NAME                  0
Rpt Dist No                0
Part 1-2                   0
Crm Cd                     0
Crime Description          0
Mocodes               151619
Vict Age                   0
Vict Sex              144644
Vict Descent          144656
Premis Cd                 16
Premis Desc              588
Weapon Used Cd        677744
Weapon Desc           677744
Status                     1
Status Desc                0
Crm Cd 1                  11
Crm Cd 2              935831
Crm Cd 3             1002677
Crm Cd 4             1004927
LOCATION                   0
Cross Street          850755
LAT                        0
LON                        0
dtype: int64

In [8]:
df = df.dropna() #drop null values
df.isnull().sum()

DR_NO                0
Date Rptd            0
DATE OCC             0
TIME OCC             0
AREA                 0
AREA NAME            0
Rpt Dist No          0
Part 1-2             0
Crm Cd               0
Crime Description    0
Mocodes              0
Vict Age             0
Vict Sex             0
Vict Descent         0
Premis Cd            0
Premis Desc          0
Weapon Used Cd       0
Weapon Desc          0
Status               0
Status Desc          0
Crm Cd 1             0
Crm Cd 2             0
Crm Cd 3             0
Crm Cd 4             0
LOCATION             0
Cross Street         0
LAT                  0
LON                  0
dtype: int64

I replaced the header "Crm Cd Desc" with "Crime Description". There was hundreds of thousands of missing values, so I removed them.

In [9]:
df['Crime Description'] #select a column 

27751                                       BRANDISH WEAPON
134755                                      BRANDISH WEAPON
363765                                       RAPE, FORCIBLE
372524                                              ROBBERY
490101    SODOMY/SEXUAL CONTACT B/W PENIS OF ONE PERS TO...
537834                                      RAPE, ATTEMPTED
586003                                           KIDNAPPING
728576                                              ROBBERY
809419                                              ROBBERY
934749                                      ORAL COPULATION
Name: Crime Description, dtype: object

In [10]:
df.sort_values(by='Crime Description') #sort data by a column

DR_NO               Date Rptd                DATE OCC  TIME OCC  \
27751   201904032  01/02/2020 12:00:00 AM  01/01/2020 12:00:00 AM      2135   
134755  200613424  08/02/2020 12:00:00 AM  08/02/2020 12:00:00 AM      2030   
586003  221401314  11/10/2022 12:00:00 AM  11/10/2022 12:00:00 AM      2117   
934749  241905348  02/04/2024 12:00:00 AM  02/03/2024 12:00:00 AM      1100   
537834  221718232  12/25/2022 12:00:00 AM  12/25/2022 12:00:00 AM      1150   
363765  210617136  10/08/2021 12:00:00 AM  10/07/2021 12:00:00 AM      1950   
372524  210209196  05/08/2021 12:00:00 AM  05/08/2021 12:00:00 AM       230   
728576  231717599  11/15/2023 12:00:00 AM  11/15/2023 12:00:00 AM       400   
809419  231915572  10/21/2023 12:00:00 AM  10/21/2023 12:00:00 AM         1   
490101  220600626  04/27/2022 12:00:00 AM  04/23/2022 12:00:00 AM      2300   

        AREA   AREA NAME  Rpt Dist No  Part 1-2  Crm Cd  \
27751     19     Mission         1924         1     761   
134755     6   Hollywood          657         1     761   
586003    14     Pacific         1452         2     910   
934749    19     Mission         1983         1     820   
537834    17  Devonshire         1797         1     122   
363765     6   Hollywood          659         1     121   
372524     2     Rampart          279         1     210   
728576    17  Devonshire         1738         1     210   
809419    19     Mission         1902         1     210   
490101     6   Hollywood          646         1     821   

                                        Crime Description  ... Status  \
27751                                     BRANDISH WEAPON  ...     AA   
134755                                    BRANDISH WEAPON  ...     AO   
586003                                         KIDNAPPING  ...     IC   
934749                                    ORAL COPULATION  ...     AO   
537834                                    RAPE, ATTEMPTED  ...     AA   
363765                                     RAPE, FORCIBLE  ...     IC   
372524                                            ROBBERY  ...     AO   
728576                                            ROBBERY  ...     IC   
809419                                            ROBBERY  ...     AA   
490101  SODOMY/SEXUAL CONTACT B/W PENIS OF ONE PERS TO...  ...     IC   

         Status Desc Crm Cd 1 Crm Cd 2  Crm Cd 3 Crm Cd 4  \
27751   Adult Arrest    761.0    930.0     997.0    998.0   
134755   Adult Other    761.0    920.0     930.0    998.0   
586003   Invest Cont    812.0    860.0     910.0    998.0   
934749   Adult Other    761.0    820.0     910.0    998.0   
537834  Adult Arrest    122.0    230.0     910.0    998.0   
363765   Invest Cont    121.0    210.0     910.0    998.0   
372524   Adult Other    210.0    510.0     910.0    998.0   
728576   Invest Cont    210.0    230.0     761.0    998.0   
809419  Adult Arrest    210.0    250.0     761.0    998.0   
490101   Invest Cont    230.0    821.0     910.0    998.0   

                               LOCATION                     Cross Street  \
27751   ASTORIA                      ST  SAN FERNANDO                 RD   
134755                          WESTERN                          ROMAINE   
586003                       WASHINGTON                         SPEEDWAY   
934749                           BURNET                        PARTHENIA   
537834  PARTHENIA                    ST                      HAYVENHURST   
363765                        NORMANDIE                       DE LONGPRE   
372524                     JAMES M WOOD                            GREEN   
728576  HASKELL                      AV  SAN FERNANDO                 BL   
809419                             POLK                           BORDEN   
490101                            SELMA                       LAS PALMAS   

            LAT       LON  
27751   34.2949 -118.4571  
134755  34.0885 -118.3092  
586003  33.9792 -118.4666  
934749  34.2282 -118.4633  
537834  34.2285 -118.4939  
363765  34.

In [11]:
df[0:5] #slice by rows

DR_NO               Date Rptd                DATE OCC  TIME OCC  \
27751   201904032  01/02/2020 12:00:00 AM  01/01/2020 12:00:00 AM      2135   
134755  200613424  08/02/2020 12:00:00 AM  08/02/2020 12:00:00 AM      2030   
363765  210617136  10/08/2021 12:00:00 AM  10/07/2021 12:00:00 AM      1950   
372524  210209196  05/08/2021 12:00:00 AM  05/08/2021 12:00:00 AM       230   
490101  220600626  04/27/2022 12:00:00 AM  04/23/2022 12:00:00 AM      2300   

        AREA  AREA NAME  Rpt Dist No  Part 1-2  Crm Cd  \
27751     19    Mission         1924         1     761   
134755     6  Hollywood          657         1     761   
363765     6  Hollywood          659         1     121   
372524     2    Rampart          279         1     210   
490101     6  Hollywood          646         1     821   

                                        Crime Description  ... Status  \
27751                                     BRANDISH WEAPON  ...     AA   
134755                                    BRANDISH WEAPON  ...     AO   
363765                                     RAPE, FORCIBLE  ...     IC   
372524                                            ROBBERY  ...     AO   
490101  SODOMY/SEXUAL CONTACT B/W PENIS OF ONE PERS TO...  ...     IC   

         Status Desc Crm Cd 1 Crm Cd 2  Crm Cd 3 Crm Cd 4  \
27751   Adult Arrest    761.0    930.0     997.0    998.0   
134755   Adult Other    761.0    920.0     930.0    998.0   
363765   Invest Cont    121.0    210.0     910.0    998.0   
372524   Adult Other    210.0    510.0     910.0    998.0   
490101   Invest Cont    230.0    821.0     910.0    998.0   

                               LOCATION                     Cross Street  \
27751   ASTORIA                      ST  SAN FERNANDO                 RD   
134755                          WESTERN                          ROMAINE   
363765                        NORMANDIE                       DE LONGPRE   
372524                     JAMES M WOOD                            GREEN   
490101                            SELMA                       LAS PALMAS   

            LAT       LON  
27751   34.2949 -118.4571  
134755  34.0885 -118.3092  
363765  34.0966 -118.3005  
372524  34.0503 -118.2720  
490101  34.0997 -118.3363  

[5 rows x 28 columns]

In [12]:
df = df[(df['Crm Cd 1']>=0) & (df['Crm Cd 2']>=0) & (df['Crm Cd 3']>=0)] #filter data
df.describe()

DR_NO     TIME OCC       AREA  Rpt Dist No   Part 1-2  \
count  1.000000e+01    10.000000  10.000000    10.000000  10.000000   
mean   2.192602e+08  1341.300000  12.500000  1303.700000   1.100000   
std    1.352419e+07   882.411979   6.720615   664.544631   0.316228   
min    2.006134e+08     1.000000   2.000000   279.000000   1.000000   
25%    2.103112e+08   575.000000   6.000000   657.500000   1.000000   
50%    2.210010e+08  1550.000000  15.500000  1595.000000   1.000000   
75%    2.292178e+08  2095.250000  18.500000  1875.750000   1.000000   
max    2.419053e+08  2300.000000  19.000000  1983.000000   2.000000   

           Crm Cd   Vict Age   Premis Cd  Weapon Used Cd    Crm Cd 1  \
count   10.000000  10.000000   10.000000       10.000000   10.000000   
mean   494.600000  33.500000  228.100000      244.600000  419.800000   
std    341.291078  19.682197  256.205842      140.814456  307.157506   
min    121.000000   0.000000  101.000000      102.000000  121.000000   
25%    210.000000  17.250000  101.000000      130.250000  210.000000   
50%    485.500000  37.500000  105.000000      208.500000  220.000000   
75%    805.250000  49.750000  120.750000      354.500000  761.000000   
max    910.000000  59.000000  718.000000      500.000000  812.000000   

         Crm Cd 2    Crm Cd 3  Crm Cd 4        LAT         LON  
count   10.000000   10.000000      10.0  10.000000   10.000000  
mean   578.100000  890.900000     998.0  34.164540 -118.402450  
std    321.082354   73.577548       0.0   0.115152    0.086572  
min    210.000000  761.000000     998.0  33.979200 -118.493900  
25%    235.000000  910.000000     998.0  34.090525 -118.465775  
50%    665.000000  910.000000     998.0  34.163950 -118.451900  
75%    850.250000  910.000000     998.0  34.259025 -118.315975  
max    930.000000  997.000000     998.0  34.310300 -118.272000

I 